lucene/core/src/java/org/apache/lucene/search/BottomValueChecker.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Maintains the bottom value across multiple collectors
+ */
+abstract class BottomValueChecker<T> {
+  /** Maintains global bottom score as the maximum of all bottom scores */
+  private static class MaximumBottomScoreChecker extends BottomValueChecker<Float> {
+    private final AtomicInteger globalBottomValue = new AtomicInteger();
+    private final AtomicBoolean bottomValueAvailable = new AtomicBoolean();
+
+    @Override
+    public void updateThreadLocalBottomValue(Float value) {
+      globalBottomValue.updateAndGet(currentValue -> Float.intBitsToFloat(currentValue) < value ? Float.floatToIntBits(value) : currentValue);
```

Comment:
Instead of using an AtomicInteger you could maybe use a double checked locking with a volatile ? Something like:
```
public void updateThreadLocalBottomValue(float value) {
   if (value <= maxMinScore) {
     return;
   }
   synchronized (this) {
       if (value > maxMinScore) {
          maxMinScore = value;
       }
  }
}
```
This would avoid the float to int conversion and should limit contention ?

Reply:
Fixed, thanks